blog/content/post/2021-02-21-tigerlake-osf.md

```diff
@@ -0,0 +1,183 @@
+there is not much one can do without support or bug history. So I have  sent an
+email to coreboot mailing list and got a reply that older microcode revisions
+had problem with new CAR setup:
+
+https://mail.coreboot.org/hyperkitty/list/coreboot@coreboot.org/thread/7YHWASZX3CQ5U3L7D5CVJCDZSMRNCCXK/
+
+I have followed the advise and turned off the INTEL_CAR_NEM_ENHANCED. I did the
+trick and I could see the serial output on the console. "Now it is a piece of
+cake" I thought... Then I saw the FSP memory init returned an error... Great,
+what now?
+
+Second issue is the memory initialization and configuration. After
+investigating the source code I noted the Tiger Lake RVP mainboard has LPDDR4
+configuration. But wait, my platform has 2 DDR4 SODIMMs... That explains the
+error. The typical different between LPDDR and DDR memory is that the former
+requires an exact mapping of memory signal from CPU to DRAM to be passed to
+FSP. So simply zeroing this configuration should be enough. Also the memory
+initialization was called by `meminit_lpddr4x` which was not suited for DDR4,
+so I have to change it to `meminit_ddr4` with appropriate parameters for SMBus
+SPD addresses. After these modifications I could past through the memory
+initialization.
+
+Another problem faced was the CPU initialization in coreboot. I was getting
+board resets during MCE (Machine Check Error registers) clearing which I have
+written about here:
+
+https://mail.coreboot.org/hyperkitty/list/coreboot@coreboot.org/thread/MW44TIEMFMVDWPVCAFPE2QUFXXGZYYAX/
+
+Till today I haven't got any response, however, I managed to resolve the
+problem as well. When investigating the logs I noticed that the microcode is
+not automatically loaded before the reset vector. For reminder, the microcode
+is being loaded through FIT table before reset vector since Haswell (4th
+generation) processors. This could be due to the processor being engineering
+sample (which is common for RVP platforms). I had placed explicit call to
+update microcode using coreboot then I got past the PCI enumeration phase.
+
+Right after the PCI enumeration phase I was hit with FSP notify error. No idea
+what could cause this issue, since the notify phases typically do not do much,
+but yet I managed to hit an error. To this point I haven't been able to figure
+out what is wrong. Trying to narrow it down with debug FSP binary didn't help
+as well, because the FSP asserts in Thunderbolt/USB type C initialization.
+
+I am still not close to booting an operating system, but yet I have encountered
+many bugs in the meantime. This is the first time I have been put through so
+much effort to boot a reference platform. Previous platform I have tried
+(Cometlake-S or Kaby Lake U) did not cause almost any problem. The quality of
+the implementation of the Intel silicon support done by Intel employees
+themselves seem to have degraded. As a proof of that please red the following
```

Comment:
@miczyg1 I would avoid this rhetoric, let's try positive feedback and encourage improvements. Let's try to not complain.

Reply:
Tried to change the post into more positive one. Please read again if you may